dipy/io/statefull_tractogram.py

```diff
@@ -0,0 +1,879 @@
+        return False
+
+    if to_space not in Space:
+        logging.error('Space MUST be one of the 3 choices (Enum)')
+        return False
+
+    if trk_header_check and extension == '.trk':
+        if not is_header_compatible(filename, reference):
+            logging.error('Trk file header does not match the provided ' +
+                          'reference')
+            return False
+
+    timer = time.time()
+    data_per_point = None
+    data_per_streamline = None
+    if extension in ['.trk', '.tck']:
+        tractogram_obj = nib.streamlines.load(filename).tractogram
+        streamlines = tractogram_obj.streamlines
+        if extension == '.trk':
+            data_per_point = tractogram_obj.data_per_point
+            data_per_streamline = tractogram_obj.data_per_streamline
+
+    elif extension in ['.vtk', '.fib']:
+        streamlines = load_vtk_streamlines(filename)
+    elif extension in ['.dpy']:
+        dpy_obj = Dpy(filename, mode='r')
+        streamlines = list(dpy_obj.read_tracks())
+    logging.debug('Load %s with %s streamlines in %s seconds',
+                  filename, len(streamlines), round(time.time() - timer, 3))
+
+    sft = StateFullTractogram(streamlines, reference, Space.RASMM,
+                              shifted_origin=shifted_origin,
+                              data_per_point=data_per_point,
+                              data_per_streamline=data_per_streamline)
+
+    if to_space == Space.VOX:
+        sft.to_vox()
+    elif to_space == Space.VOXMM:
+        sft.to_voxmm()
+
+    if bbox_valid_check and not sft.is_bbox_in_vox_valid():
+        raise ValueError('Bounding box is not valid in voxel space, cannot ' +
+                         'load a valid file if some coordinates are invalid')
+
+    return sft
+
+
+def robust_split_name(filename):
+    """ Get the basename and extension of a filename, robust to nii.gz """
+    base, ext = os.path.splitext(filename)
+    if ext == ".gz":
+        temp_base, add_ext = os.path.splitext(base)
+
+        if add_ext == ".nii":
+            ext = add_ext + ext
+            base = temp_base
+
+    return base, ext
+
+
+def get_reference_info(reference):
+    """ Will compare the spatial attribute of 2 references
+
+    Parameters
+    ----------
+    reference : Nifti or Trk filename, Nifti1Image or TrkFile, Nifti1Header or
+        trk.header (dict)
+        Reference that provides the spatial attribute.
+
+    Returns
+    -------
+    output : tuple
+        - affine ndarray (4,4), np.float32, tranformation of VOX to RASMM
+        - dimensions list (3), int, volume shape for each axis
+        - voxel_sizes  list (3), float, size of voxel for each axis
+        - voxel_order, string, Typically 'RAS' or 'LPS'
+    """
+
+    is_nifti = False
+    is_trk = False
+    if isinstance(reference, six.string_types):
+        _, ext = robust_split_name(reference)
+        if ext == '.nii' or ext == '.nii.gz':
```

Comment:
Personally, I would avoid checking for specific extensions and delegate that to Nibabel. I'd try to do a `nib.load` (or if it fails a `nib.streamlines.load`) and check the type of object that is returned (like you do anyway). I think that makes it more generic since you could provide a BytesIO object and makes it more robust to people messing with extension.

Reply:
And then you would continue with the check for instance of Nifti1Image, TrkFile, Nifti1Header, etc. ?

Reply:
Yes.